### ISSUE TITLE
Using  Configuration Endpoint to Connect AWS Elasticcache for REDIS

### ISSUE THREAD

**Issue by pthargis** [StackExchange/StackExchange.Redis#2237]
I have an Elasticache instance running in cluster mode with one shard with three nodes (1 primary, 2 replicas) in different AZs.

Our Redis Client code uses StackExchange REDIS Library to connect AWS Elasticache for REDIS. We use the AWS Elasticache configuration endpoint (not the specific node endpoint) while connecting it. According to the AWS Documentation, the Configuration Endpoint will be updated with the new primary after the failover is completed. Here are my questions 

Does the StackExchange REDIS Library leverage AWS Elasticache Configuration Endpoint to handle the routing of the requests to the right Primary?

Or 

Does StackExcahge REDIS library auto-discover (using INFO commands) the primary  & secondary endpoints and make a connection to the correct Primary? How does it behave after the failover?

Can you please help us to understand? 

@NickCraver @mgravell

**Comment by NickCraver** [comment]
As far as I know, their failover is purely DNS as far as we're concerned, so we'll just re-resolve on the next connection. Offer stands for the AWS crew to support something like server maintenance notifications Azure does for proactively doing this as far as possible, but we haven't heard back on that front. The current version of the client will buffer and re-send some connections though.

Be sure you _aren't_ using `resolveDns=true` for your connection, as it needs to re-resolve on each connection for the AWS mechanism to work.

**Comment by pthargis** [comment]
Thank you, @NickCraver . Curious to know Why the StackExchange REDIS library run INFO commands on all the nodes? Is it for Heartbeat check or for auto discovery?

**Comment by NickCraver** [comment]
Yeah sure - to see what we're connected to. Are we writable? Which version? Things like that. We're specifically looking at the replication and server sections only, if on a version new enough to support `INFO <section>`

**Comment by pthargis** [comment]
Thank you

**Comment by pthargis** [comment]
@NickCraver - When we use ConnectionMultiplexer, Does it create a connection per node/shard if it is Azure or AWS elasticate? For example, if we have one primary and two replicas, does it maintain three active connections totally?

**Comment by pthargis** [comment]
@NickCraver  - One more question. When I perform a simple Set command, I see SE Redis Library creates two connections (standalone Redis).

The cmd block indicates below.
cmd=setex
cmd=subscribe

See the full details. May I know the purpose of the 2nd connection cmd=subscribe ? 

id=582 addr=127.0.0.1:33090 fd=9 name= age=9 idle=8 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=0 obl=0 oll=0 omem=0 tot-mem=20520 events=r cmd=setex user=default
id=583 addr=127.0.0.1:33092 fd=10 name= age=9 idle=8 flags=P db=0 sub=1 psub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=0 obl=0 oll=0 omem=0 tot-mem=20504 events=r cmd=subscribe user=default

**Comment by NickCraver** [comment]
Yeah sure - with Redis subscription connections have to be separate (listening to change events, etc.) - it's a separation in the RESP 2 protocol. RESP3 can multiplex here but it's not in wide use anywhere we're finding and so we haven't added support in the library either.

On the total count: we create 2 per endpoint 1 "interactive" (everything but subscriptions) and 1 for subscriptions.
